### PR TITLE
Cleanup script to remove a collection created by a plugin.

### PR DESCRIPTION
**Pull request by Natim**
- [x] Add documentation.
- [x] Add tests.
- [x] Add a changelog entry.

### CODE REVIEW

**Comment by Natim** [comment]
I wonder if we want to add this to the `kinto` command itself. In that case it would be great to add permission removal as well.

**Comment by leplatrem** [comment]
Related https://github.com/mozilla-services/cliquet/issues/583 *«that'll come back later through the CLI and kinto-admin»* 😄

File: scripts/cleanup.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from __future__ import print_function
+import argparse
+import warnings
+import sys
+from pyramid.paster import bootstrap
+from pyramid.settings import asbool
+
+
+CONFIG_FILE = 'config/kinto.ini'
+DEFAULT_COLLECTION = '/buckets/blocklists/collections/certificates,'
```
Comment (leplatrem) [inline scripts/cleanup.py]: hum, no :)

File: scripts/cleanup.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    parser = argparse.ArgumentParser(description="Kinto commands")
+    parser.add_argument('--ini',
+                        help='Application configuration file',
+                        dest='ini_file',
+                        required=False,
+                        default=CONFIG_FILE)
+
+    parser.add_argument('--collection',
+                        help='Collection to cleanup',
```
Comment (leplatrem) [inline scripts/cleanup.py]: Maybe not «cleanup», but delete records ?

File: scripts/cleanup.py (hunk 1 above)
Comment (Natim) [inline scripts/cleanup.py]: For now that's the purpose of this script.

File: scripts/cleanup.py (hunk 2 above)
Comment (Natim) [inline scripts/cleanup.py]: It delete the collection not only its records.

**Comment by leplatrem** [comment]
It could be `kinto delete-collection --bucket="bid" collection="cid"` :)

**Comment by Natim** [comment]
Where should I put the documentation?

**Comment by leplatrem** [comment]
We can start a page for the kinto CLI ? Don't we have the output of `--help` somewhere already ?

File: kinto/__main__.py [hunk 3]
```diff
@@ -46,6 +47,17 @@ def main(args=None):
     parser_migrate = subparsers.add_parser('migrate')
     parser_migrate.set_defaults(which='migrate')
 
+    parser_delete_collection = subparsers.add_parser('delete-collection')
+    parser_delete_collection.add_argument(
+        '--bucket',
+        help='The bucket the collection is in.',
```
Comment (leplatrem) [inline kinto/__main__.py]: @glasserc how would you say that in colloquial english?

File: kinto/core/scripts.py [hunk 4]
```diff
@@ -25,3 +24,37 @@ def migrate(env):
...
+
+
+def delete_collection(env, bucket_id, collection_id):
+    registry = env['registry']
+    settings = registry.settings
+    readonly_mode = asbool(settings.get('readonly', False))
+
+    if readonly_mode:
+        message = ('Cannot delete the collection while in readonly mode.')
+        warnings.warn(message)
```
Comment (leplatrem) [inline kinto/core/scripts.py]: genuine question: Are warnings always shown, no matter how python was compiled/ran ?

File: kinto/core/scripts.py [hunk 5]
```diff
@@ -25,3 +24,37 @@ def migrate(env):
...
+        return 31
+
+    collection = '/buckets/%s/collections/%s' % (bucket_id, collection_id)
```
Comment (leplatrem) [inline kinto/core/scripts.py]: What happens if you call instance_uri with `Request.blank()` ?

**Comment by leplatrem** [comment]
f+ :)

File: kinto/core/scripts.py (hunk 4 above)
Comment (Natim) [inline kinto/core/scripts.py]: Would you rather print an error message to the stderr?

File: kinto/core/scripts.py (hunk 5 above)
Comment (leplatrem) [inline kinto/core/scripts.py]: Here you go:
```
(.venv) ➜  kinto git:(master) ✗ python
Python 2.7.12 (default, Jul  1 2016, 15:12:24) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from webob.request import Request
>>> from kinto.core.utils import instance_uri
>>> r = Request.blank(path="/")
>>> instance_uri(r, "bucket", id="toto")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "kinto/core/utils.py", line 397, in instance_uri
    return strip_uri_prefix(request.route_path('%s-record' % resource_name,
  File "/home/mathieu/Code/Mozilla/kinto/.venv/local/lib/python2.7/site-packages/webob/request.py", line 1395, in __getattr__
    raise AttributeError(attr)
AttributeError: route_path

```

Comment (Natim) [inline kinto/core/scripts.py]: Does it means it doesn't work? :)

**Comment by leplatrem** [comment]
@Natim final r+ ?

capture d ecran de 2016-07-28 17-58-22
capture d ecran de 2016-07-28 17-58-13

File: kinto/__main__.py (hunk 3 above)
Comment (glasserc) [inline kinto/__main__.py]: We discussed this on IRC. "The bucket the collection is in" sounds fine to me, but if you don't like it, some other alternatives include "The bucket where the collection is", "The bucket with the collection in it", "The bucket containing the collection", or "The bucket that contains the collection".

File: kinto/__main__.py [hunk 6]
```diff
@@ -98,16 +121,18 @@ def main(args=None):
...
         dry_run = parsed_args['dry_run']
-        try:
-            logging.config.fileConfig(config_file)
-        except configparser.NoSectionError as e:  # pragma: no cover
-            print(e)
         env = bootstrap(config_file)
         scripts.migrate(env, dry_run=dry_run)
 
-    elif parsed_args['which'] == 'start':
+    elif parsed_args['which'] == 'delete-collection':
```
Comment (Natim) [inline kinto/__main__.py]: Shall we use `which_command` here too?

File: kinto/core/scripts.py (hunk 4 above)
Comment (Natim) [inline kinto/core/scripts.py]: I guess you solved this with your logger configuration?

File: kinto/core/scripts.py [hunk 7]
```diff
@@ -25,3 +27,62 @@ def migrate(env, dry_run=False):
...
+
+    bucket = '/buckets/%s' % bucket_id
+    collection = '/buckets/%s/collections/%s' % (bucket_id, collection_id)
+
+    try:
+        registry.storage.get(collection_id='bucket',
+                             parent_id='',
+                             object_id=bucket_id)
+    except storage_exceptions.RecordNotFoundError:
+        print("Bucket %r does not exist." % bucket)
```
Comment (Natim) [inline kinto/core/scripts.py]: Can we use logger here then.

File: kinto/core/scripts.py [hunk 8]
```diff
@@ -25,3 +27,62 @@ def migrate(env, dry_run=False):
...
+        return 33
+
+    try:
+        registry.storage.get(collection_id='collection',
+                             parent_id=bucket,
+                             object_id=collection_id)
+    except storage_exceptions.RecordNotFoundError:
+        print("Collection %r does not exist." % collection)
```
Comment (Natim) [inline kinto/core/scripts.py]: ditto

File: kinto/core/scripts.py [hunk 9]
```diff
@@ -25,3 +27,62 @@ def migrate(env, dry_run=False):
...
+    bucket = '/buckets/%s' % bucket_id
+    collection = '/buckets/%s/collections/%s' % (bucket_id, collection_id)
+
+    try:
+        registry.storage.get(collection_id='bucket',
+                             parent_id='',
+                             object_id=bucket_id)
+    except storage_exceptions.RecordNotFoundError:
+        print("Bucket %r does not exist." % bucket)
+        return 33
```
Comment (Natim) [inline kinto/core/scripts.py]: We should use 32 here since 33 is already used for collection.

File: kinto/core/scripts.py [hunk 10]
```diff
@@ -25,3 +27,62 @@ def migrate(env, dry_run=False):
...
+                             object_id=collection_id)
+    except storage_exceptions.RecordNotFoundError:
+        print("Collection %r does not exist." % collection)
+        return 33
+
+    deleted = registry.storage.delete_all(collection_id='record',
+                                          parent_id=collection,
+                                          with_deleted=False)
+    if deleted == 0:
+        print('No records found for %r.' % collection)
```
Comment (Natim) [inline kinto/core/scripts.py]: Exchange print by a logger call.

**Comment by Natim** [comment]
Test are failing because we are misisng a bit of coverage:

    kinto/core/scripts.py                             47      7    85%   49-51, 57-59, 65

**Comment by Natim** [comment]
Final r? @leplatrem

File: kinto/core/__init__.py [hunk 11]
```diff
@@ -173,5 +173,5 @@ def add_api_capability(config, identifier, description="", url="", **kw):
     config.scan("kinto.core.views")
 
     # Give sign of life.
-    msg = "%(project_name)s %(project_version)s starting."
+    msg = "Working with %(project_name)s %(project_version)s."
```
Comment (leplatrem) [inline kinto/core/__init__.py]: Not ideal... «Loading ?»

Comment (Natim) [inline kinto/core/__init__.py]: Or maybe nothing?

Comment (Natim) [inline kinto/core/__init__.py]: Running ?

File: docs/commandline.rst [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    Objects are permanently deleted, and there is no way to cancel the operation.
+
+::
+
+    usage: kinto delete-collection [-h] --bucket BUCKET --collection COLLECTION
+
+    optional arguments:
+      -h, --help            show this help message and exit
+      --bucket BUCKET       The bucket where the collection is.
```
Comment (Natim) [inline docs/commandline.rst]: where the collection stands?

File: kinto/__main__.py [hunk 13]
```diff
@@ -4,73 +4,93 @@
...
 from pyramid.scripts import pserve
 from pyramid.paster import bootstrap
 from kinto import __version__
 from kinto.config import init
 
-CONFIG_FILE = 'config/kinto.ini'
+DEFAULT_CONFIG_FILE = 'config/kinto.ini'
+DEFAULT_PORT = 8888
+DEFAULT_LOG_LEVEL = logging.INFO
+DEFAULT_LOG_FORMAT = "%(levelname)-5.5s  %(message)s"
```
Comment (leplatrem) [inline kinto/__main__.py]: Why did you remove the time? It can be useful if some actions are ran in jobs?

File: kinto/__main__.py [hunk 14]
```diff
@@ -98,16 +118,18 @@ def main(args=None):
                 import pip
                 pip.main(['install', "kinto[postgresql]"])
 
-    elif parsed_args['which'] == 'migrate':
+    elif which_command == 'migrate':
         dry_run = parsed_args['dry_run']
-        try:
-            logging.config.fileConfig(config_file)
-        except configparser.NoSectionError as e:  # pragma: no cover
-            print(e)
```
Comment (leplatrem) [inline kinto/__main__.py]: I think we should keep the logging from .ini. For example, it allows to turn-on debug during migrations

File: kinto/__main__.py (hunk 13 above)
Comment (Natim) [inline kinto/__main__.py]: Usually jobs runner takes care of this. This command is meant to be run by hand and it wanted to have the most usuful output.

File: kinto/__main__.py (hunk 14 above)
Comment (Natim) [inline kinto/__main__.py]: I think we should add [verbose options](https://github.com/Kinto/kinto-http.py/blob/master/kinto_http/cli_utils.py#L65-L76) then, but the logger config of the server  with the HekaLogger doesn't fit well with running this command.

**Comment by leplatrem** [comment]
r+
